### PR TITLE
Rust lib improvements

### PR DESCRIPTION
**Pull request by Esgrove**
- Add a few helper methods and structs to make code less noise and easier to follow
- Move lib functions only used in vault to associated functions inside vaul implementation
- Add `new` for `CloudFormationParams` that takes direct ownership of parameters
- Delete s3 objects with one call using `delete_objects`

### CODE REVIEW

File: rust/src/lib.rs [hunk 1]
```diff
@@ -65,44 +77,74 @@ impl CloudFormationParams {
...
-        let vault_key_from_env = get_env_variable("VAULT_KEY");
+        let vault_stack_from_env = Self::get_env_variable("VAULT_STACK");
+        let vault_bucket_from_env = Self::get_env_variable("VAULT_BUCKET");
+        let vault_key_from_env = Self::get_env_variable("VAULT_KEY");
 
         let cloudformation_params = match (vault_bucket_from_env, vault_key_from_env) {
-            (Some(bucket), Some(key)) => {
-                CloudFormationParams::from(bucket.as_str(), Some(key.as_str()))
-            }
+            (Some(bucket), Some(key)) => CloudFormationParams::new(bucket, Some(key)),
```
Comment (Esgrove) [inline rust/src/lib.rs:140]: bucket and key can be moved to `CloudFormationParams` since they are not used after this

File: rust/src/lib.rs [hunk 2]
```diff
@@ -70,39 +83,83 @@ impl fmt::Display for CloudFormationParams {
...
-        // Check env variables directly in case library is not used through the CLI.
-        // These are also handled in the CLI so they are documented in the CLI help.
-        let vault_stack_from_env = get_env_variable("VAULT_STACK");
-        let vault_bucket_from_env = get_env_variable("VAULT_BUCKET");
-        let vault_key_from_env = get_env_variable("VAULT_KEY");
+        // Check env variables directly in case the library is not used through the CLI.
+        // These are also handled in the CLI, so they are documented in the CLI help.
+        let vault_stack_from_env = Self::get_env_variable("VAULT_STACK");
+        let vault_bucket_from_env = Self::get_env_variable("VAULT_BUCKET");
+        let vault_key_from_env = Self::get_env_variable("VAULT_KEY");
```
Comment (kahlstrm) [inline rust/src/lib.rs:152]: Why encapsulate helper functions inside the `impl Vault`. This makes no sense as those are unrelated to `Vault`. Also involves adding this `Self::`to everything, which just doesn't do anything.

File: rust/src/lib.rs [hunk 3]
```diff
@@ -70,39 +83,83 @@ impl fmt::Display for CloudFormationParams {
...
-            }
+            (Some(bucket), Some(key)) => CloudFormationParams::new(bucket, Some(key)),
             (_, _) => {
                 let stack_name = vault_stack_from_env
                     .as_deref()
                     .or(vault_stack)
                     .unwrap_or("vault");
-                get_cloudformation_params(&config, stack_name).await?
+
+                Self::get_cloudformation_params(&config, stack_name).await?
```
Comment (kahlstrm) [inline rust/src/lib.rs:162]: This should be in `impl CloudFormationParams` rather than in `impl Vault`

File: rust/src/lib.rs [hunk 4]
```diff
@@ -119,7 +176,7 @@ impl Vault {
         region_opt: Option<&str>,
     ) -> Result<Vault, VaultError> {
         let config = aws_config::from_env()
-            .region(get_region_provider(region_opt))
+            .region(Self::get_region_provider(region_opt))
```
Comment (kahlstrm) [inline rust/src/lib.rs:179]: Don't need to encapsulate this as it doesn't use `Self` or create one.

File: rust/src/lib.rs [hunk 5]
```diff
@@ -70,39 +83,83 @@ impl fmt::Display for CloudFormationParams {
...
+}
+
 impl Vault {
     pub async fn new(
         vault_stack: Option<&str>,
         region_opt: Option<&str>,
     ) -> Result<Vault, VaultError> {
         let config = aws_config::from_env()
-            .region(get_region_provider(region_opt))
+            .region(Self::get_region_provider(region_opt))
```
Comment (kahlstrm) [inline rust/src/lib.rs:144]: Don't move this inside `impl Vault`.

File: rust/src/lib.rs [hunk 6]
```diff
@@ -300,92 +354,92 @@ impl Vault {
...
-
-    Ok(CloudFormationParams {
-        bucket_name: parse_output_value_from_key("vaultBucketName", stack_output)
-            .ok_or(VaultError::BucketNameMissingError)?,
-        key_arn: parse_output_value_from_key("kmsKeyArn", stack_output),
-        // deployed_version: parse_output_value_from_key("vaultStackVersion", &stack_output),
-    })
-}
+    /// Get CloudFormation parameters based on config and stack name
+    async fn get_cloudformation_params(
```
Comment (kahlstrm) [inline rust/src/lib.rs:398]: This should be in `impl CloudFormationParams`

File: rust/src/lib.rs [hunk 7]
```diff
@@ -300,92 +354,92 @@ impl Vault {
...
-fn parse_output_value_from_key(key: &str, out: &[Output]) -> Option<String> {
-    out.iter()
-        .find(|output| output.output_key() == Some(key))
-        .map(|output| output.output_value().unwrap_or_default().to_owned())
-}
+    fn parse_output_value_from_key(key: &str, out: &[Output]) -> Option<String> {
+        out.iter()
+            .find(|output| output.output_key() == Some(key))
+            .map(|output| output.output_value().unwrap_or_default().to_owned())
+    }
```
Comment (kahlstrm) [inline rust/src/lib.rs:432]: This should be in `impl CloudFormationParams` if it were to be moved anywhere

File: rust/src/lib.rs [hunk 8]
```diff
@@ -300,92 +354,92 @@ impl Vault {
...
+        nonce
+    }
 
-/// Return possible env variable value as Option
-fn get_env_variable(name: &str) -> Option<String> {
-    env::var(name).ok()
+    /// Return possible env variable value as Option
+    fn get_env_variable(name: &str) -> Option<String> {
+        env::var(name).ok()
+    }
```
Comment (kahlstrm) [inline rust/src/lib.rs:444]: I'm not sure why we even have this function, as `env::var(name).ok()` should suffice on its own. Nevertheless, should not be in `impl Vault`

**Review by kahlstrm — CHANGES_REQUESTED** [review]
Again I'm not seeing any big improvements made here, but I guess. Would appreciate if one were to look into #336 #246 #216 instead, as those would be more major improvements.

File: rust/src/lib.rs (hunk 2 above)
Comment (Esgrove) [reply rust/src/lib.rs:152]: These are functions that are only used inside Vault. Rust calls these _associated_ functions, other languages like C++, C#, and Python etc calls them  _static_ functions. The point is to put them inside since they are related to the Vault in that they are only used by it. 

https://doc.rust-lang.org/reference/items/associated-items.html#associated-functions-and-methods

Comment (Esgrove) [reply rust/src/lib.rs:152]: `Self::` is needed to reference these functions since they do not operate on an instance of the type, similarly to `new()`.

File: rust/src/lib.rs (hunk 8 above)
Comment (Esgrove) [reply rust/src/lib.rs:444]: Why not, it is only used there and nowhere else. 

ChatGPT:

> Associated functions help in organizing code through namespacing. By associating functions with the types they operate on or are related to, Rust allows for cleaner code organization and easier navigation. It becomes immediately clear which type a function is meant to be used with, which is especially helpful in larger codebases or when using multiple modules.

Comment (Esgrove) [reply rust/src/lib.rs:444]: And why have `get_env_variable` in the first place: it is much more descriptive and hides the actual implementation for getting the env variable, which we are not interested in. It is immediately clear what that does making the code more readable.

File: rust/src/lib.rs (hunk 4 above)
Comment (Esgrove) [reply rust/src/lib.rs:179]: Self is here as the namespace, it is not about using anything from the instance or creating one.

File: rust/src/lib.rs [hunk 9]
```diff
@@ -70,39 +83,83 @@ impl fmt::Display for CloudFormationParams {
...
+
+    /// Convert keys to S3 object identifiers.
+    pub fn to_object_identifiers(&self) -> Vec<ObjectIdentifier> {
+        self.as_array()
+            .iter()
+            .map(|key| {
+                ObjectIdentifier::builder()
+                    .set_key(Some(key.to_string()))
+                    .build()
+                    .unwrap_or_else(|_| panic!("Failed to create ObjectIdentifier for '{key}'"))
```
[the remark this answers is not in the document]
Comment (Esgrove) [reply rust/src/lib.rs]: Changed 👍🏻

**Review by kahlstrm — APPROVED** [review]
